### PR TITLE
WIP: Improve Handler behavior names

### PR DESCRIPTION
**Pull request by Head0nF1re**
PR for issue #8:

- Add snapshot tests for identifier names.

- Add property `Behavior.RegistrationTypeAsIdentifer` computed from `Behavior.RegistrationType` in order to generate a valid identifier name. Generates an identifier based on the fully qualified name of the type, preventing potential issues where types with the same name are on different namespaces.

- Change Handler scriban template (**failing**)

### CODE REVIEW

**Review by viceroypenguin — CHANGES_REQUESTED** [review]
Two comments (and a way to fix them):
* Adding a mutable field to the `Behavior` type means that it is no longer properly equatable, which means it is no longer correctly cacheable. 
* The rendered names are way too long.

Instead of adding something to the `Behavior` type, we can do something slightly different.

1. Update the `Behavior` type to receive the simple name of the type, which can be obtained in TransformBehaviors by calling `.Name` of the originaldefinition. 
2. Update transform handlers code in the entrypoint file to use something like:
```cs
var names = new Dictionary<string, int>()
{
  ["HandleBehavior"] = 1,
};

string GetSuffix(string name)
{
  if (!names.TryGetValue(name, out var cnt))
    cnt = 0;
  names[name] = cnt + 1;
  return cnt == 0 ? "" : $"{cnt}";
}

var renderBehaviors = behaviors
  .Select(b => new
  {
    b.NonGenericTypeName,
    VariableName = b.Name + GetSuffix(b.Name),
  })
  .ToList();
```

3. Then pass the new `renderBehaviors` to the template instead, and it can get the variable from the `{{ b.behavior_name }}` value.

**Comment by Head0nF1re** [comment]
@viceroypenguin 

> The rendered names are way too long.

Yeah, thought it would be more readable in case a user wants to peek into the generated `Handler`.

With **fully qualified name**:
```csharp
_HandleBehavior = HandleBehavior;
_Dummy_LoggingBehavior = Dummy_LoggingBehavior;
_YetAnotherDummy_OtherBehavior = YetAnotherDummy_OtherBehavior;
_Dummy_SecondLoggingBehavior = Dummy_SecondLoggingBehavior;
_YetAnotherDummy_LoggingBehavior = YetAnotherDummy_LoggingBehavior;
_YetAnotherDummy_SecondLoggingBehavior = YetAnotherDummy_SecondLoggingBehavior;

_YetAnotherDummy_SecondLoggingBehavior.SetInnerHandler(_HandleBehavior);
_YetAnotherDummy_LoggingBehavior.SetInnerHandler(_YetAnotherDummy_SecondLoggingBehavior);
_Dummy_SecondLoggingBehavior.SetInnerHandler(_YetAnotherDummy_LoggingBehavior);
_YetAnotherDummy_OtherBehavior.SetInnerHandler(_Dummy_SecondLoggingBehavior);
_Dummy_LoggingBehavior.SetInnerHandler(_YetAnotherDummy_OtherBehavior);
``` 

With **class name**:
```csharp
_HandleBehavior = HandleBehavior;
_LoggingBehavior1 = LoggingBehavior1;
_OtherBehavior = OtherBehavior;
_SecondLoggingBehavior1 = SecondLoggingBehavior1;
_LoggingBehavior2 = LoggingBehavior2;
_SecondLoggingBehavior2 = SecondLoggingBehavior2;

_SecondLoggingBehavior2.SetInnerHandler(_HandleBehavior);
LoggingBehavior2.SetInnerHandler(_SecondLoggingBehavior2);
SecondLoggingBehavior1.SetInnerHandler(_LoggingBehavior2);
OtherBehavior.SetInnerHandler(_SecondLoggingBehavior1);
LoggingBehavior1.SetInnerHandler(_OtherBehavior);
``` 

This would be more meaningful in the case where the same `Type.Name` would be declared in different namespaces like we talked on Discord, since the type information isn't displayed on hover you would need to scroll up and down (see the var types) to better visualize what's going on. 
But yeah, since that will rarely happen, guess that's not a problem. Will update then.


---

Regarding the caching, you mean in the [Incremental Generator](https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md#caching)? Help me visualize please, doesn't the generator run `TransformBehaviors` each time since the predicate only checks if the node is `CompilationUnitSyntax`? Or, would the problem be in later stages of the pipeline?

File: src/Immediate.Handlers.Generators/ImmediateHandlers/ImmediateHandlersGenerator_TransformBehaviors.cs [hunk 1]
```diff
@@ -87,6 +87,7 @@ CancellationToken cancellationToken
 					NonGenericTypeName = constructorType,
 					RequestType = constraint.RequestType,
 					ResponseType = constraint.ResponseType,
+					Name = originalDefinition.Name
```
Comment (viceroypenguin) [inline src/Immediate.Handlers.Generators/ImmediateHandlers/ImmediateHandlersGenerator_TransformBehaviors.cs:90]: add a comma here

File: src/Immediate.Handlers.Generators/ImmediateHandlers/ImmediateHandlersGenerator_TransformHandler.cs [hunk 2]
```diff
@@ -88,16 +88,16 @@ CancellationToken cancellationToken
 
 	private static RenderMode? GetOverrideRenderMode(INamedTypeSymbol symbol) =>
 		symbol.GetAttribute("Immediate.Handlers.Shared.RenderModeAttribute")
-				is { } rma
```
Comment (viceroypenguin) [inline src/Immediate.Handlers.Generators/ImmediateHandlers/ImmediateHandlersGenerator_TransformHandler.cs:91]: unexpected whitespace changes in this file

File: src/Immediate.Handlers.Generators/ImmediateHandlers/ImmediateHandlersGenerator_Entrypoint.cs [hunk 3]
```diff
@@ -146,6 +146,28 @@ Template template
...
+		string GetVariableNameSuffix(string typeName)
+		{
+			if (!typesCount.TryGetValue(typeName, out var count))
+				count = 0;
+
+			typesCount[typeName] = count + 1;
+			return count == 0 ? string.Empty : $"{count}";
+		}
+
+		var renderBehaviors = pipelineBehaviors.Select(b =>
```
Comment (viceroypenguin) [inline src/Immediate.Handlers.Generators/ImmediateHandlers/ImmediateHandlersGenerator_Entrypoint.cs:163]: for readability, change to:
```cs
var renderBehaviors = pipelineBehaviors
  .Select(..
  )
  .ToList();
```

**Review by viceroypenguin — CHANGES_REQUESTED** [review]

**Pull request merged**